File: PERL/Cap08/c08ex01.pl
# c08ex01.pl

#!/usr/bin/perl

use strict;
use warnings;

my $NUMERO;

inicio:
  print "Entre um valor numerico inteiro: ";
  $NUMERO = <STDIN>;
  chomp($NUMERO);
  print "\n";

  if ($NUMERO ne "") {
    if (caractereEspecial($NUMERO)) {
      print "Erro: Caractere especial nao permitido.\n\n";
    } elsif (validaInteiro($NUMERO)) {
      goto fim;
    } else {
      print "Erro: Entre um numero inteiro.\n\n";
    }
  } else {
      print "Erro: Tecla <Enter> acionada ";
      print "acidentalmente.\n\n";
  }
goto inicio;
fim:

print "Entrada valida fornecida: $NUMERO\n";

print "\n";
print "Tecle <Enter> para encerrar... ";
<STDIN>;

sub caractereEspecial {
  my ($ENTRADA) = @_;
  my $CARACTERE = "!\@#\$%^&*()-_=+[]{}|;:',<>/?`~\"";
    
  for my $i (0..length($CARACTERE)-1) {
    if (index($ENTRADA, substr($CARACTERE, $i, 1)) > -1) {
      return 1;
    }
  }
    
  return 0;
}

sub validaInteiro {
  my ($ENTRADA) = @_;
  return $ENTRADA =~ /^[+-]?\d+$/;
}

File: C/Cap08/c08ex01.c
// c08ex01.c

#include <stdio.h>
#include <stdbool.h>
#include <ctype.h>
#include <string.h>

bool validaInteiro(const char* Numero) {
  int i;
  int tamanho = strlen(Numero);

  if (tamanho > 0 && Numero[0] == '-')
    i++;

  for (i = 0; i <= tamanho - 1; ++i) {
    if (!(isdigit(Numero[i])))
      return false;
  }

  return true;
}

int main(void) {
  char Numero[100];

  do {
    printf("Entre um valor numerico inteiro: ");
    fgets(Numero, sizeof(Numero), stdin);
    Numero[strcspn(Numero, "\n")] = '\0';
    printf("\n");


    if (Numero[0] == '\0') {
      printf("Erro: Tecla <Enter> acionada acidentalmente.\n");
      printf("\n");
      continue;
    }

    if (!validaInteiro(Numero)) {
      printf("Erro: Entre um numero inteiro.\n");
      printf("\n");
      continue;
    }

    break;
  } while (true);

  printf("Entrada valida fornecida: %s\n", Numero);

  printf("\n");
  printf("Tecle <Enter> para encerrar... ");
  while (getchar() != '\n');

  return 0;
}
